### PR TITLE
Wrap git_merge_file

### PR DESCRIPTION
**Pull request by alehed**
Add a wrapper for git_merge_file

### CODE REVIEW

**Comment by alehed** [comment]
Just some changes I had lying around.

File: ObjectiveGit/GTRepository+Merging.h [hunk 1]
```diff
@@ -52,6 +53,18 @@ typedef NS_OPTIONS(NSInteger, GTMergeAnalysis) {
...
+///
+/// The parameters taked are the ones received from `enumerateConflictedFiles`.
+///
+/// ancestor - The ancestor entry
+/// ours     - The index entry of our side
+/// theirs   - The index entry of their side
+/// error    - The error if one occurred. Can be NULL.
+///
+/// Returns The file content annotated with conflict markers or null on error
+- (NSString* _Nullable)stringForConflictWithAncestor:(GTIndexEntry *)ancestor ourSide:(GTIndexEntry *)ourSide theirSide:(GTIndexEntry *)theirSide withError:(NSError **)error;
```
Comment (tiennou) [inline ObjectiveGit/GTRepository+Merging.h]: Small nitpicks :
- `*` should be right-aligned
- `withError:` => `error:`

File: ObjectiveGit/GTRepository+Merging.m [hunk 2]
```diff
@@ -170,6 +172,86 @@ - (BOOL)mergeBranchIntoCurrentBranch:(GTBranch *)branch withError:(NSError **)er
...
+	if (theirData == nil) {
+		return nil;
+	}
+	NSString *theirString = [[NSString alloc] initWithData: theirData encoding:NSUTF8StringEncoding];
+	theirInput.ptr = [theirString cStringUsingEncoding:NSUTF8StringEncoding];
+	theirInput.size = theirString.length;
+
+
+	git_merge_file_result result;
+	git_merge_file(&result, &ancestorInput, &ourInput, &theirInput, nil);
```
Comment (tiennou) [inline ObjectiveGit/GTRepository+Merging.m]: Missing error checking.

File: ObjectiveGit/GTRepository+Merging.h (hunk 1 above)
Comment (tiennou) [inline ObjectiveGit/GTRepository+Merging.h]: There's a few more `*` spacing weirdness in the hunk below as well.

**Comment by tiennou** [comment]
Thanks for the PR ! A few more thoughts though :
- I'd say using `NSString` will prevents that from working with binary files. Would it be acceptable to switch to using `NSData` inside the method, just in case ?
- I find the `-stringForConflictWithAncestor:…` name unclear. May I suggest something like `-contentsOfDiffWithAncestor:…` or something along those lines ?

**Comment by alehed** [comment]
@tiennou thanks for the feedback! I'll fix it this weekend.

**Comment by alehed** [comment]
Addressed the main issues. The test is pretty messy, but I don't see a nicer or shorter way to produce conflicting commits on two different branches and merging them.

**Comment by alehed** [comment]
This should be ready.

File: ObjectiveGit/GTRepository+Merging.m [hunk 3]
```diff
@@ -170,6 +172,87 @@ - (BOOL)mergeBranchIntoCurrentBranch:(GTBranch *)branch withError:(NSError **)er
...
+	if (gitError != GIT_OK) {
+		if (error != NULL) *error = [NSError git_errorFor:gitError description:@"Failed to create merge file"];
+		return nil;
+	}
+
+	char *cString = malloc(result.len * sizeof(char *) + 1);
+	strncpy(cString, result.ptr, result.len);
+	cString[result.len] = '\0';
+
+	NSString *mergedContent = [[NSString alloc] initWithCString:cString encoding:NSUTF8StringEncoding];
```
Comment (tiennou) [inline ObjectiveGit/GTRepository+Merging.m]: Maybe `-initWithBytesNoCopy:length:encoding:freeWhenDone:` ? That should allow you to just pass `result.ptr`/`result.len`, skipping the need for the `malloc`/`strncpy` step. You would have to `result.ptr = NULL` though, so `git_merge_file_result_free` doesn't break it by freeing before returning.

**Comment by tiennou** [comment]
After a more thorough look at what `libgit2` exposes, I have a feeling that it might be a too brittle wrapper around the `git_merge_file_result` [structure](https://libgit2.github.com/libgit2/#HEAD/type/git_merge_file_result). But I'll defer to @pietbrauer on that though.

**Comment by alehed** [comment]
I could wrap `git_merge_file_result` in a class if you mean that. But I don't know where else it would be used.

**Comment by pietbrauer** [comment]
I'm ok with not having a wrapper for now.

**Comment by tiennou** [comment]
Closing in favor of #664. Thanks for the PR @alehed, and sorry for the long turnaround!

**Pull request closed without merging**